Enable 3DS verification during checkout

v18 opened the pull request:
Closes #90

### Summary of changes

3DS support was added using the manual confirmation flow:

https://stripe.com/docs/payments/payment-intents/quickstart#manual-confirmation-flow

The whole checkout process looks like this now:

1. Customer checks out ('Place Order') 
2. WooCommerce triggers the handler for the event `checkout_place_order_woocommerce_payments` ([ref](https://github.com/woocommerce/woocommerce/blob/307096178304d62b8f927f27c64b375986192744/assets/js/frontend/checkout.js#L443)). If the handler returns false, then the check out does not proceed. This is what happens while we wait for a response from Stripe about whether or not the form requires additional verification in our event handler.
3. After we get back a response from Stripe, it either needs extra verification or not. If it does, we ask the user for it using `handleCardAction` ([ref](https://github.com/Automattic/woocommerce-payments/compare/Automattic:f9ad486...Automattic:e839f73#diff-ca240674e46858808b4eecceae9d70ccR74))
4. After any extra verification is done, the form is re-submitted ([ref](https://github.com/Automattic/woocommerce-payments/compare/Automattic:f9ad486...Automattic:e839f73#diff-ca240674e46858808b4eecceae9d70ccR83))
5. This time our event handler does not return false so the checkout goes through, and `process_payment` is called.

#### Still to do, part of this PR:

- [x] make sure that cart is blocked when submitting payment, before processing payment
- [x] security: escape variables, etc

#### To do, not part of this PR:

- [x] handle response errors from Stripe in `start_payment_process`
- [x] handle failed verification
- [x] handle response errors from Stripe after `process_payment`
- add unit tests for js - #159 
- add unit tests for `deserialize_intention_object_from_array` et al
- handle the case where the payment fails and the user does not refresh the page before re-submitting the form - #160

### Testing instructions

1. Add some products to cart
2. Check out using WCS with a test credit card requiring verification, [a list is here](https://stripe.com/docs/payments/3d-secure#three-ds-cards)

v18 committed:
Add client secret to Intention model

v18 committed:
Use manual confirmation when creating intention

v18 committed:
Create intention before submitting cart

This is needed to allow further customer
action on the front end before submitting payment,
for example for 3DS.

RadoslavGeorgiev reviewed (changes requested):
I love the idea of creating the intent before the order, but without automatic confirmation. Why did you not do this before/while we were coding 4.2.0? :p

I'm not sure where we took the wrong turn with the Stripe SCA release, but what you are doing here makes a ton of sense. We were afraid to create the intent before the actual submission of the checkout form, because at that point we thought we __have to__ also authenticate the charge. In this PR, you are simply authrozing the intent, but neither authenticating nor capturing the charge. This makes all the sense in the world.

At the moment this looks like a very good approach and I cannot think of any reason why it wouldn't work. Please check the changes that I requested, and I'll re-review the PR and re-test it to see if there are some incompatible edge-case scenarios or something like that.

v18 committed:
Use $intention name in fn deserializing array

v18 committed:
Submit cart when req'd customer actions are done

If further action was required by the customer,
for example for 3DS, use `handleCardAction` before
submitting the cart.

Requires the payment intent ID.

v18 committed:
Checkout: block form before creating intention

v18 committed:
Use nonce with create_payment_intention

v18 committed:
Show default error message on payment failure

Check for error.userErrorMessage instead of the error.message since
JS errors will have the message param, and we may not want to display them
to the user. This way we are sure we intend to display the
error message to the user.

RadoslavGeorgiev reviewed (approved):
There is one last detail left to iron out: https://github.com/Automattic/woocommerce-payments/pull/133#discussion_r309654724. If you are keen on leaving it this way, go ahead, we will probably need to modify it anyway when it comes time to update intents.

Question: Why are you force-pushing so often? Every time I try to review the PR and pull, I get conflicts.

dechov reviewed (commented):
Left 2 small comments, by no means blockers for merge.
